elide-core/src/main/java/com/yahoo/elide/graphql/PersistentResourceFetcher.java

```diff
@@ -79,6 +117,151 @@ private void logContext(RelationshipOp operation, Environment environment) {
+     * @param offset pagination offset argument
+     * @param first pagination first argument
+     * @param filters filter params
+     * @return list of persistent resource objects
+     */
+    private Object fetchObject(RequestScope requestScope, Class recordType, Optional<List<String>> ids,
+                               GraphQLType outputType, Optional<String> sort, Optional<String> offset,
+                               Optional<String> first, Optional<String> filters) {
+        /* fetching a collection */
+        if(!ids.isPresent()) {
+            List<PersistentResource> records = new ArrayList<>(PersistentResource.
+                    loadRecords(recordType, requestScope, Optional.empty()));
+
+            //TODO: paginate/filter/sort
+            return records;
+
+        } else { /* fetching by id(s) */
+            List<String> idList = ids.get();
+
+            /* handle empty list of ids */
+            if(idList.isEmpty()) {
+                throw new BadRequestException("Empty list passed to ids");
+            }
+
+            /* access records from internal db and return */
+            Optional<FilterExpression> filterExpression;
+            EntityDictionary dictionary = requestScope.getDictionary();
+
+            Class<?> idType = dictionary.getIdType(recordType);
+
+            String idField = dictionary.getIdFieldName(recordType);
+
+            String entityTypeName = dictionary.getJsonAliasFor(recordType);
+
+            /* construct a new SQL like filter expression, eg: book.id IN [1,2] */
+            filterExpression = Optional.of(new FilterPredicate(
+                    new FilterPredicate.PathElement(
+                            recordType,
+                            entityTypeName,
+                            idType,
+                            idField),
+                    Operator.IN,
+                    new ArrayList<>(idList)));
+
+            return PersistentResource.loadRecords(recordType, requestScope, filterExpression);
+        }
+    }
+
+    /**
+     * FETCH attributes of top level entity
+     * @param parentResource parent object
+     * @param field Field type
+     * @return list of persistent resource objects
+     */
+    private Object fetchAttribute(PersistentResource parentResource, Field field) {
+        return parentResource.getAttribute(field.getName());
+    }
+
+    /**
+     * Fetches a relationship for a top-level entity
+     * @param context environment encapsulating graphQL's request environment
+     * @return persistence resource object(s)
+     */
+    private Object fetchRelationship(Environment context) {
+        EntityDictionary dictionary = context.requestScope.getDictionary();
+        Class parentClass = context.parentResource.getResourceClass();
+        String fieldName = context.field.getName();
+        String ID = "id";
+
+        /* check for toOne relationships */
+        if(dictionary.getRelationshipType(parentClass, fieldName).isToOne()) {
```

Comment:
You don't need to differentiate the persistentResource call based on the relationshipType.

There is a single method which you should call for both toMany and toOne relationships.  That call will always return a Set.

What you can do instead is look at the relationship type to determine whether to return the single instance from that set or to return the set.